### DISCUSSION TITLE
{cli} output not captured by quarto slides

### DISCUSSION THREAD

**Discussion by JosiahParry** [quarto-dev/quarto-cli#3957] [Q&A]
Output from the cli package is not captured by reveal js slides. 

Desired behavior: output from cli is captured and printed 

Reprex:

````
---
title: "this-dont-work"
format: revealjs
---

## define class with cli print

```{r, echo = TRUE}
print.myclass <- function(x, ...) cli::cli_text(cli::style_bold("this shouldnt work"))

obj <- structure(list(1, 2, 3), class = "myclass")

obj
```

## change the print method

```{r, echo = TRUE}
print.myclass <- function(x, ...) cat("this will work")

obj
```

````

**Comment by jthomasmock** [comment]
Try this from Danielle Navarro maybe?

https://blog.djnavarro.net/posts/2021-04-18_pretty-little-clis/

**Reply by JosiahParry** [reply]
Unfortunately this doesn't work in quarto slides (at least the way im going about it).

````
---
title: "this-dont-work"
format: revealjs
---

```{r}
options(crayon.enabled = TRUE)
ansi_aware_handler <- function(x, options) {
  paste0(
    "<pre class=\"r-output\"><code>",
    fansi::sgr_to_html(x = x, warn = FALSE, term.cap = "256"),
    "</code></pre>"
  )
}

knitr::knit_hooks$set(
  output = ansi_aware_handler, 
  message = ansi_aware_handler, 
  warning = ansi_aware_handler,
  error = ansi_aware_handler
)

```


## define class with cli print

```{r, echo = TRUE}
print.myclass <- function(x, ...) cli::cli_text(cli::style_bold("this shouldnt work"))

obj <- structure(list(1, 2, 3), class = "myclass")

obj
```

## change the print method

```{r, echo = TRUE}
print.myclass <- function(x, ...) cat("this will work")

obj
```
````

**Comment by mcanouil** [comment]
I am not completely sure to fully grasp the extent of what you want to achieve, but you should write a `knit_print` S3 method and register it, see <https://bookdown.org/yihui/rmarkdown-cookbook/opts-render.html#opts-render>.

````markdown
---
format: revealjs
---

```{r}
#| echo: true
#| results: asis
options(crayon.enabled = TRUE)
library(knitr)
knit_print.myclass <- function(x, options) {
  fansi::sgr_to_html(x = cli::style_bold("this should'nt work"), warn = FALSE, term.cap = "256")
}
registerS3method("knit_print", "myclass", knit_print.myclass)

(obj <- structure(list(1, 2, 3), class = "myclass"))

print(obj)
```
````

<img width="728" alt="image" src="https://user-images.githubusercontent.com/8896044/212179843-4666a7b8-fd12-4c97-8a88-1afddee28c4f.png">

**Reply by JosiahParry** [reply]
Are you suggesting that for an any object that uses {cli} in its print method there should also be a knit_print method?

**Reply by JosiahParry** [reply]
The goal is to capture the output of various objects that use {cli} in their print method, so I think the above would be a little much

**Reply by mcanouil** [reply]
The issue is the call to `cli::cli_text()`, see an example below:

````
---
format: revealjs
---

```{r}
#| results: asis
#| echo: false
old_hook <- fansi::set_knit_hooks(knitr::knit_hooks, which = c("output", "warning", "error", "message"))
options(crayon.enabled = TRUE)
```

## {.smaller}

- Does not work because of `cli::cli_text()`

```{r}
#| echo: true
print.myclass <- function(x, ...) {
  cli::cli_text(cli::style_bold(cli::col_blue("this "), cli::col_red("shouldnt "), " work ..."))
  1 + 1
}

(obj <- structure(list(1, 2, 3), class = "myclass"))
```

- Does ot work because nothing is printing the cli message.

```{r}
#| echo: true
print.myclass <- function(x, ...) {
  cli::style_bold(cli::col_blue("this "), cli::col_red("shouldnt "), "work ...")
  1 + 1
}
obj
```

- Does work properly.

```{r}
#| echo: true
print.myclass <- function(x, ...) {
  cat(cli::style_bold(cli::col_blue("this "), cli::col_red("shouldnt "), "work ..."))
  1 + 1
}
obj
```
````

<img width="715" alt="image" src="https://user-images.githubusercontent.com/8896044/212185269-65af737c-c651-4a31-b4fb-d1c207d81ab8.png">

**Reply by mcanouil** [reply]
I am not sure there is a workaround to the following behaviour ...
<img width="336" alt="image" src="https://user-images.githubusercontent.com/8896044/212187344-a2dafa1b-af5f-4523-b03f-d94c34362222.png">

unfortunately, I am not familiar with `cli` internals.

**Reply by JosiahParry** [reply]
real odd! oh well :/ im not sure if it is worth a {cli} bug report or not

**Reply by mcanouil** [reply]
I do not think it is a bug.

Let's ping/ask @gaborcsardi on this 😇

**Comment by cderv** [comment]
@JosiahParry if I understand correctly you would like to have CLI output like colors showed in a knitted output ? 

**cli** is working with ANSI coloring and is made for terminal. Neither in R Markdown or Quarto you would get the formatting. I don't think it is a revealjs issue - it will not show any formatting in any format by default. Historically it was a **crayon** issue https://github.com/r-lib/crayon/issues/24#issuecomment-581068792

To get ANSI formatting in HTML output, it would require specific handling, which the **fansi** package used by @mcanouil in his answer is doing (See vignette https://cran.r-project.org/web/packages/fansi/vignettes/sgr-in-rmd.html). 
Or some tools like **pkgdown** and **downlit** are specifically converting ANSI character for formatting to HTML formatting. Quarto do not by default and integration with **downlit** needs to be improve (see https://github.com/quarto-dev/quarto-cli/issues/3730)

Regarding why printing with `cli::cli_text()` output shows in `format: html` and not `format: revealjs`, this is because `cli::cli_text()` and probably other **cli** functions are like `message()` or `warning()`. So set `message: true` or `warning: true` and you will see them, as revealjs sets `warning` and `message` (and `echo`) to FALSE by default 

````markdown
---
title: "this-dont-work"
format: revealjs
---

## define class with cli print

```{r}
#| echo: true
#| message: true
print.myclass <- function(x, ...) cli::cli_text(cli::style_bold("this shouldnt work"))

obj <- structure(list(1, 2, 3), class = "myclass")

obj
```

## change the print method

```{r}
#| echo: true
print.myclass <- function(x, ...) cat("this will work")

obj
```
````
Probably `cli::cat_*()` functions are more comparable to `cat()` and don't require to activate `message: true` in revealjs. 

Hope it helps understand.

**Reply by mcanouil** [reply]
@cderv 🤦‍♂️ I did not thought of looking at the default of `message`, `warning`, and `error`.
I did not realise the default was different from `html` format.

**Reply by cderv** [reply]
No problem. You can't know everything :) 

They are different because space in slide is rare so by default it is better to limit the outputs to only results. Hence the different execute options.

**Reply by JosiahParry** [reply]
@cderv im not worried about colored output. It would just be nice to have _any_ output. The output format for github_document in Rmarkdown renders {cli} fine enough for my standards :) it's surprising to me that there is _nothing_ though for revealjs

**Reply by JosiahParry** [reply]
I'll try setting echo and message in the chunks and see how behavior changes

**Reply by mcanouil** [reply]
@JosiahParry to have all output messages from cli, the following yaml header will do the job. 😉

```yaml
execute:
  message: true
  warning: true
  error: true
```

**Comment by t-kalinowski** [comment]
I wonder if Quarto should/could automatically do the equivalent for R users when it knows the final output format is html:

````r
```{r setup, include = FALSE}
options(cli.num_colors = 256)
fansi::set_knit_hooks(knitr::knit_hooks)
```
````

Something like this would "just work" for most users.

The equivalent functionality could alternatively be done as post-processing step, if injecting knitr hooks before evaluating code is not desirable. (This might require an update in `cli::num_ansi_colors()` to make it aware of Quarto's new capability to handle ansi color sequences). In pseudo code:
```r
if(fansi::has_sgr(x))
  x <- fansi::sgr_to_html(fansi::html_esc(x))
```

**Reply by cderv** [reply]
Thanks Tomasz ! Are you suggesting to use **fansi** package logic for knitr engine only, or reimplement the logic in quarto cross-language so that ANSI are handled in HTML output  ? 

Just to be sure to understand. 

We could set a hook for **knitr** to use, but if so we don't want to depend on the ANSI package in Quarto. 

However, is this is only done in R side of Quarto, this could probably also be done directly in **rmarkdown** as an opt-in or else.

**Comment by t-kalinowski** [comment]
I think ideally, this would be implemented directly in quarto, so we could get cross-language support from one effort. However, I understand development time is limited, and `fansi` in R already exists and does most of the hard work.  

My pragmatic suggestion is to inject `fansi` hooks when using the R knitr engine, with the understanding that ansi sequence will only "just work" for users using the knitr engine.

**Reply by gaborcsardi** [reply]
FWIW cli can convert ANSI tags to HTML, see `?cli::ansi_html`. I am pretty sure that there are JS/TS packages that can do this as well, if you want to have in the quarto.
